test/adapters/UniswapAdapter.js

```diff
@@ -0,0 +1,361 @@
+            BigNumber.from("1000000000000000000000000000000000"),
+            { from: user }
+          );
+
+          const promise = this.adapter.buyLp(
+            "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
+            BigNumber.from("10000000000000"),
+            this.exchangeName,
+            0,
+            0,
+            {
+              from: user,
+            }
+          );
+          await expect(promise).to.be.revertedWith("not enough funds");
+        });
+
+        it("valid purchase with eth", async function () {
+          const promise = this.adapter.buyLp(
+            "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
+            this.ethAmt,
+            this.exchangeName,
+            0,
+            0,
+            {
+              from: user,
+              value: this.ethAmt,
+            }
+          );
+        });
+
+        it("valid purchase with wbtc", async function () {
+          const btc_amt = await this.token.balanceOf(user, { from: user });
+
+          await this.weth.deposit({
+            from: user,
+            value: BigNumber.from("100000000000000000"),
+          });
+          await this.weth.approve(
+            this.uniswap.address,
+            BigNumber.from("1000000000000000000000000000000000"),
+            { from: user }
+          );
+          await this.uniswap.swapExactTokensForTokens(
+            BigNumber.from("1000000000000000"),
+            0,
+            [
+              "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
+              "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
+            ],
+            user,
+            BigNumber.from("10000000000000000000000000000000"),
+            { from: user }
+          );
+          const approval = await this.token.approve(
+            this.adapter.address,
+            BigNumber.from("1000000000000000000000000000000000"),
+            { from: user }
+          );
+          const promise = this.adapter.buyLp(
+            "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
+            this.wbtcAmt,
+            this.exchangeName,
+            0,
+            0,
+            {
+              from: user,
+            }
+          );
+        });
+
+        it("purchase with eth using expected amounts out from adapter", async function () {
+          const expWbtc = await this.adapter.expectedWbtcOut(
+            this.ethAmt,
+            this.exchangeName,
+            {
+              from: user,
+            }
+          );
+          const expDigg = await this.adapter.expectedDiggOut(
+            expWbtc,
+            this.exchangeName,
+            {
+              from: user,
+            }
+          );
+          const promise = this.adapter.buyLp(
+            "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
+            this.ethAmt,
+            this.exchangeName,
+            expWbtc - 1,
+            expDigg - 1,
+            {
+              from: user,
+              value: this.ethAmt,
+            }
+          );

```

Comment:
Let's test a few states after each test:

- LP share balance of user
- WBTC/Digg balance increase of Uniswap pool